File: lc/coreprep/implementation/practice_iterations/patterns_iterations_1/k_closest_points_to_origin.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <list>
#include <queue>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

struct Point {
  int x;
  int y;

  Point() : x(-1), y(-1) {}

  Point(int x, int y) : x(x), y(y) {}
};

double euclidean_distance(const Point &p) {
  return sqrt((double)p.x * p.x + (double)p.y * p.y);
}

struct CompPoint {
  const vector<Point> &a;

  CompPoint(const vector<Point> &a) : a(a) {}

  bool operator()(const Point &p1, const Point &p2) {
    return euclidean_distance(p1) < euclidean_distance(p2);
  }
};

// O(nlogk) / O(k)
vector<Point> k_closest_points_to_origin(const vector<Point> &a, int k) {
  CompPoint comp(a);
  priority_queue<Point, vector<Point>, CompPoint> max_heap(comp);

  for (int i = 0; i < k; i++) {
    max_heap.push(a[i]);
  }

  for (int i = k; i < a.size(); i++) {
    if (!max_heap.empty() &&
        euclidean_distance(a[i]) < euclidean_distance(max_heap.top())) {
      max_heap.pop();
      max_heap.push(a[i]);
    }
  }

  vector<Point> res;
  while (!max_heap.empty()) {
    res.push_back(max_heap.top());
    max_heap.pop();
  }

  return res;
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, k, x, y;
  cin >> tt;

  while (tt--) {
    cin >> n >> k;

    vector<Point> a(n);
    for (int i = 0; i < n; i++) {
      cin >> x >> y;
      a[i] = Point(x, y);
    }

    auto res = k_closest_points_to_origin(a, k);
    for (auto p : res) {
      cout << p.x << " " << p.y << "\n";
    }
  }

  return 0;
}

File: lc/coreprep/implementation/practice_iterations/patterns_iterations_2/KClosestPointsToOrigin.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KClosestPointsToOrigin {
  static class Point {
    int x;
    int y;

    Point(int x, int y) {
      this.x = x;
      this.y = y;
    }
  }

  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int k = in.nextInt();

        Point[] a = new Point[n];
        for (int i = 0; i < n; i++) {
          int x = in.nextInt();
          int y = in.nextInt();
          a[i] = new Point(x, y);
        }

        var res = kClosestPointsToOrigin(a, k);
        for (var r : res) {
          System.out.printf("%d %d\n", r.x, r.y);
        }
      }
    }
  }

  private static double euclideanDistance(Point p) {
    return Math.sqrt(p.x * p.x + p.y * p.y);
  }

  // O(nlogk) / O(k)
  public static List<Point> kClosestPointsToOrigin(Point[] a, int k) {
    PriorityQueue<Integer> maxHeap =
        new PriorityQueue<>((x, y)
                                -> Double.compare(euclideanDistance(a[y]),
                                                  euclideanDistance(a[x])));

    for (int i = 0; i < k; i++) {
      maxHeap.add(i);
    }

    for (int i = k; i < a.length; i++) {
      if (euclideanDistance(a[i]) < euclideanDistance(a[maxHeap.peek()])) {
        maxHeap.poll();
        maxHeap.add(i);
      }
    }

    List<Point> res = new ArrayList<>();
    while (!maxHeap.isEmpty()) {
      res.add(a[maxHeap.poll()]);
    }

    return res;
  }
}